File: objects.py
# Core

# Function value
class Function(object):
	def __init__(self, body, parameters, parents):
		self.body = body
		self._items = parameters
		self._parents = parents
		self._context = {}

	def call(self, arguments):
		# This time, it's actually being called. Map args and pass to body
		# TODO: Default args, possible named params
		if len(arguments) != len(self._items):
			raise TypeError() # TODO: more info

		arguments = dict(zip(map(lambda i: i.name, self._items), arguments))
		arguments.update(self._context)

		func = self.get_body()
		return func.call(arguments)

	def get_body(self):
		# What am i doing even i don't know any more.
		return Body(self.body, self._parents)

# Seriously hacky shit used by nodes to fake a scope
class Namespace(object):
	def __init__(self, name, child):
		self._name = name
		self._child = child

	def get(self, name, limiter=''):
		if name == self._name:
			return self._child
		return Null()

	def set(self, name, value, limiter=''):
		# Can't override a namespace, soz
		raise AttributeError() # TODO: more info

# Object formed by a called function
class Body(object):
	def __init__(self, expressions=None, parents=None):
		if expressions is None:
			expressions = []
		if parents is None:
			parents = []

		self._items = expressions
		self._parents = parents
		self._context = {}

	def add_parent(self, parent):
		self._parents.append(parent)

	def add_context(self, context):
		self._context.update(context)

	def call(self, arguments=None):
		# Arguments (if passed) will be {name: value} that should be added to context
		if arguments:
			self._context.update(arguments)

		for expression in self._items:
			expression.execute(self)
		return self
	__call__ = call

	def get(self, name, limiter=''):
		if 'self' in limiter:
			if name in self._context:
				return self._context[name]

		if 'parent' in limiter:
			for parent in self._parents:
				result = parent.get(name, limiter)
				if result:
					return result

		# Nothing found, null it
		return Null()

	def set(self, name, value, limiter=''):
		# Check local scope first
		if 'self' in limiter and name in self._context:
			self._context[name] = value
			return

		# Wasn't local, check parents
		if 'parent' in limiter:
			for parent in self._parents:
				if parent.get(name, limiter):
					parent.set(name, value, limiter)
					return

		# Wasn't in a parent either, set new var locally
		if 'self' in limiter:
			self._context[name] = value
			return

		# Limiter excluding self, but nothing found. Raise error.
		raise AttributeError() # TODO: more info


# Acts as a context, and value. Absorbs all sets and returns itself when retrieved.
class Null(object):
	def evaluate(self, scope):
		return self

	def assign(self, value, scope):
		pass

	def get(self, name, limiter=''):
		return self

	def set(self, name, value, limiter=''):
		pass

	def string(self, indent=0):
		return  ' '*indent+'(null)';

	def __bool__(self):
		return False

File: object/objects.go
package object

// Function

type Function struct {
	expressions []Expression
	parameters  []string

	parents  []Context
	defaults map[string]Object
}

func NewFunction(parameters []string, expressions []Expression) *Function {
	function := new(Function)

	function.parameters = parameters
	function.expressions = expressions

	return function
}

func (f *Function) Call(arguments []Object, scope Context) Context {
	return f.CallWithContext(f.createContext(arguments, scope))
}

func (f *Function) createContext(arguments []Object, scope Context) Context {
	// TODO: Throw error if length of args is != length of params (default args/named params?)

	// Copying, so that injected arguments do not taint next call
	defaults := make(map[string]Object)
	for k, v := range f.defaults {
		defaults[k] = v
	}

	for i, parameter := range f.parameters {
		defaults[parameter] = arguments[i]
	}

	// TODO: Add scope to temporary parents list

	return &Dictionary{defaults, f.parents}
}

func (f *Function) CallWithContext(context Context) Context {
	for _, expression := range f.expressions {
		expression.Execute(context)
	}

	return context
}

func (f *Function) Type() string {
	return "function"
}

// Dictionary

type Dictionary struct {
	values  map[string]Object
	parents []Context
}

func (d *Dictionary) Get(name string) Object {
	// TODO: only if local scope allowed
	if value, ok := d.values[name]; ok {
		return value
	}

	// TODO: only if parent scope allowed
	// TODO: probably need to change the conditional down the line
	for _, parent := range d.parents {
		result := parent.Get(name)
		if result != nil {
			return result
		}
	}

	return new(Null)
}

func (d *Dictionary) Set(name string, value Object) {
	// Check local scope
	// TODO: only if local scope allowed
	if _, ok := d.values[name]; ok {
		d.values[name] = value
	}

	// It wasn't local, check parents
	// TODO: only is parent scope is allowed
	// TODO: probably need to change the conditional down the line
	for _, parent := range d.parents {
		if parent.Get(name) != nil {
			parent.Set(name, value)
			return
		}
	}

	// Wasn't in parent either, set new var locally
	// TODO: only if local scope allowed
	d.values[name] = value

	// TODO: when limiting, if it gets this far, throw error
}

func (d *Dictionary) Type() string {
	return "dictionary"
}

// Namespace

type Namespace struct {
	Name string
	Value Object
}

func (n *Namespace) Get(name string) Object {
	if name == n.Name {
		return n.Value
	}
	return new(Null)
}

func (n *Namespace) Set(name string, value Object) {
	// Should not be able to set on a namespace, throw error
}

func (n *Namespace) Type() string {
	return "namespace"
}

// Null

type Null struct{}

func (n *Null) Get(name string) Object {
	return n
}

func (m *Null) Set(name string, value Object) {
	// Setting a value on null does nothing
}

func (n *Null) Type() string {
	return "null"
}

// Number

type Number struct {
	Value float64
}

func (n *Number) Type() string {
	return "number"
}

// String

type String struct {
	Value string
}

func (s *String) Type() string {
	return "string"
}
